include/boost/url/detail/impl/remove_dot_segments.ipp

```diff
@@ -0,0 +1,496 @@
+//
+// Copyright (c) 2022 alandefreitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_DETAIL_IMPL_REMOVE_DOT_SEGMENTS_IPP
+#define BOOST_URL_DETAIL_IMPL_REMOVE_DOT_SEGMENTS_IPP
+
+#include <boost/url/detail/remove_dot_segments.hpp>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+std::size_t
+remove_dot_segments(
+    char* dest0,
+    char const* end,
+    string_view s,
+    bool remove_unmatched) noexcept
+{
+    // 1. The input buffer is initialized with
+    // the now-appended path components and the
+    // output buffer is initialized to the empty
+    // string.
+    char* dest = dest0;
+    auto append = [&dest0, &dest, &end]
+        (string_view in) -> std::size_t
+    {
+        if (in.size() > std::size_t(end - dest))
+        {
+            dest = dest0;
+            return false;
+        }
+        std::memcpy(dest, in.data(), in.size());
+        dest += in.size();
+        return true;
+    };
+
+    auto find_last_slash =
+        [&dest0, &dest]() -> std::size_t
+    {
+        char* const first = dest0;
+        char* last = dest;
+        if (first == last)
+            return string_view::npos;
+        do {
+          --last;
+          if (*last == '/')
+                return last - first;
+        } while (last != first);
+        return string_view::npos;
+    };
+
+    // Step 2 is a loop through 5 production rules
+    // There are no transitions between all rules,
+    // which enables some optimizations.
+    // A.  If the input buffer begins with a
+    // prefix of "../" or "./", then remove
+    // that prefix from the input buffer;
+    // otherwise,
+    // Rule A can only happen at the beginning:
+    // - B and C write "/" to the output
+    // - D can only happen at the end
+    // - E leaves "/" or happens at the end
+    while (!s.empty())
+    {
+        if (s.starts_with("../"))
+        {
+            if (!remove_unmatched)
+            {
+                if (!append(s.substr(0, 3)))
+                {
+                    return 0;
+                }
+            }
+            s.remove_prefix(3);
+            continue;
+        }
+        if (s.starts_with("./"))
+        {
+            s.remove_prefix(2);
+            continue;
+        }
+        break;
```

Comment:
You can reorder these statements to remove a line:
```
...
    if( ! s.starts_with("./") )
        break;
    s.remove_prefix(2);
}
```

There are multiple benefits here. It is easier to read since it loops again where people expect. The conditional is expressed so that the most likely outcome evaluates to true (some optimizers care about this). And it is less code.